### PR TITLE
vmselect: add support of multi-tenant queries

### PR DESCRIPTION
**Pull request by zekker6**
### Describe Your Changes

Added an ability to query data across multiple tenants. See: https://github.com/VictoriaMetrics/VictoriaMetrics/issues/1434

Currently, the following endpoints work with multi-tenancy:
- /prometheus/api/v1/query
- /prometheus/api/v1/query_range
- /prometheus/api/v1/series
- /prometheus/api/v1/labels
- /prometheus/api/v1/label/<label_name>/values
- /prometheus/api/v1/status/active_queries
- /prometheus/api/v1/status/top_queries
- /prometheus/api/v1/status/tsdb
- /prometheus/api/v1/export
- /prometheus/api/v1/export/csv
- /vmui

A note regarding VMUI: endpoints such as `active_queries` and `top_queries` have been updated to indicate whether query was a single-tenant or multi-tenant, but UI needs to be updated to display this info.
cc: @Loori-R 

### Checklist

The following checks are **mandatory**:

- [x] My change adheres [VictoriaMetrics contributing guidelines](https://docs.victoriametrics.com/contributing/).

### CODE REVIEW

File: app/vmselect/multitenant/filters.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+	"strings"
+
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/auth"
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/prompbmarshal"
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/promrelabel"
+)
+
+// ApplyFiltersToTenants applies the given filters to the given tenants.
+// It returns the filtered tenants.
+func ApplyFiltersToTenants(tenants, filters []string) ([]*auth.Token, error) {
```
Comment (f41gh7) [inline app/vmselect/multitenant/filters.go:14]: I think, behavior must be consistent with vminsert - if neither vm_account_id nor  vm_project_id is defined for filters - execute request to the 0:0 tenant (fast path)

File: app/vmselect/prometheus/prometheus.go [hunk 2]
```diff
@@ -695,16 +843,35 @@ func LabelsHandler(qt *querytracer.Tracer, startTime time.Time, at *auth.Token,
...
 	}
 	denyPartialResponse := httputils.GetDenyPartialResponse(r)
-	sq := storage.NewSearchQuery(at.AccountID, at.ProjectID, cp.start, cp.end, cp.filterss, *maxLabelsAPISeries)
-	labels, isPartial, err := netstorage.LabelNames(qt, denyPartialResponse, sq, limit, cp.deadline)
-	if err != nil {
-		return fmt.Errorf("cannot obtain labels: %w", err)
+
+	rc := getResultsCollector()
+	defer putResultsCollector(rc)
+	if err = executeAcrossTenants(qt, at, cp, func(qt *querytracer.Tracer, at *auth.Token, _ bool) error {
```
Comment (f41gh7) [inline app/vmselect/prometheus/prometheus.go:849]: It's better to move executeAcrossTenants to netstorage level and add tenant related labelsl to the blocks parsing.

 It allows to aggregate metrics over tenants properly.

File: app/vmselect/promql/eval_test.go [hunk 3]
```diff
@@ -170,3 +173,54 @@ func TestGetSumInstantValues(t *testing.T) {
...
+	f(`up{a="b",vm_account_id="1",vm_project_id="2" or vm_account_id="3"}`, `up{a="b"}`, []string{`{vm_account_id="1",vm_project_id="2"}`, `{vm_account_id="3"}`})
+
+	f(`rate(foo{a="b"}[5m])`, `rate(foo{a="b"}[5m])`, nil)
+	f(`rate(foo{vm_account_id="1",a="b"}[5m])`, `rate(foo{a="b"}[5m])`, []string{`{vm_account_id="1"}`})
+	f(`sum(rate(foo{vm_account_id="1",a="b"}[5m]))`, `sum(rate(foo{a="b"}[5m]))`, []string{`{vm_account_id="1"}`})
+
+	f(`sum(rate(foo{vm_account_id="1",a="b"}[5m]))`, `sum(rate(foo{a="b"}[5m]))`, []string{`{vm_account_id="1"}`})
+	f(`sum(rate(sum(rate(foo{vm_account_id="1",a="b"}[5m]))))`, `sum(rate(sum(rate(foo{a="b"}[5m]))))`, []string{`{vm_account_id="1"}`})
+
+	f(`sum_over_time(rate(foo{vm_account_id="1",a="b"}[5m])[5m:1m])`, `sum_over_time(rate(foo{a="b"}[5m])[5m:1m])`, []string{`{vm_account_id="1"}`})
```
Comment (f41gh7) [inline app/vmselect/promql/eval_test.go:224]: it'd be great to have a test for the following expression: 

```
up{vm_account_id="5"} + up{vm_project_id="11"} + sum(rate(up{vm_account_id="22"}))
```

**Review by pmatuszak — APPROVED** [review]

**Review by pmatuszak — APPROVED** [review]

**Comment by FOWind** [comment]
it's unable to eval number with this pr
image

**Comment by zekker6** [comment]
@FOWind thank you for sharing, I'm currently working on quite large re-write for this PR and will make sure to check this use case as well.

**Comment by zekker6** [comment]
@FOWind This PR have been updated with a new implementation which also fixes an issue with number evaluation you've shared. You can find a pre-built image to run some tests here - [victoriametrics/vmselect:heads-multitenant-queries-0-gfdbd484883](https://hub.docker.com/layers/victoriametrics/vmselect/heads-multitenant-queries-0-gfdbd484883/images/sha256-9723cdef1916021851fb3322b1b893f12b286d47ee44356a1bc180f6d3ae7f16?context=explore)

If you'll encounter any other issues please post a comment here and we'll get it fixed, thank you for the feedback!

**Comment by zekker6** [comment]
You can find a demo datasource which is using version of vmselect build from this PR [here](https://play-grafana.victoriametrics.com/explore?schemaVersion=1&panes=%7B%22pcw%22:%7B%22datasource%22:%22P8FF7EDDF703E5E58%22,%22queries%22:%5B%7B%22refId%22:%22A%22,%22expr%22:%22up%22,%22range%22:true,%22instant%22:true,%22datasource%22:%7B%22type%22:%22prometheus%22,%22uid%22:%22P8FF7EDDF703E5E58%22%7D,%22editorMode%22:%22code%22,%22legendFormat%22:%22__auto%22%7D%5D,%22range%22:%7B%22from%22:%22now-1h%22,%22to%22:%22now%22%7D%7D%7D&orgId=1&chunkNotFound).

Feel free to try out queries and/or grouping of data across tenants. Feedback is much appreciated.

File: app/vmselect/prometheus/prometheus.go [hunk 4]
```diff
@@ -901,6 +938,11 @@ func QueryHandler(qt *querytracer.Tracer, startTime time.Time, at *auth.Token, w
 		DenyPartialResponse: httputils.GetDenyPartialResponse(r),
 		QueryStats:          qs,
 	}
+	err = populateAuthTokens(qt, ec, at, deadline)
```
Comment (f41gh7) [inline app/vmselect/prometheus/prometheus.go:941]: It's better to parse tenant related labels only once, during query request to the storage. 

 It ll make code simpler and allow to use current cache implementation.

File: lib/storage/search.go [hunk 5]
```diff
@@ -296,6 +299,33 @@ func (sq *SearchQuery) GetTimeRange() TimeRange {
 	}
 }
 
+// Exec calls cb for with marshaled requestData for each tenant in sq.
+func (sq *SearchQuery) Exec(qt *querytracer.Tracer, cb func(qt *querytracer.Tracer, requestData []byte, t TenantToken) any) []any {
```
Comment (f41gh7) [inline lib/storage/search.go:303]: It's better to move `Exec` function to the app/vmselect/netstorage package. 

It should accept search query as argument.

**Review by f41gh7 — CHANGES_REQUESTED** [review]
please see comments

File: app/vmselect/netstorage/tenant_filters.go [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/storage"
+)
+
+// GetTenantTokensFromFilters returns the list of tenant tokens and the list of filters without tenant filters.
+func GetTenantTokensFromFilters(qt *querytracer.Tracer, tr storage.TimeRange, tfs [][]storage.TagFilter, deadline searchutils.Deadline) ([]storage.TenantToken, [][]storage.TagFilter, error) {
+	tenants, err := TenantsCached(qt, tr, deadline)
+	if err != nil {
+		return nil, nil, fmt.Errorf("cannot obtain tenants: %w", err)
+	}
+
```
Comment (f41gh7) [inline app/vmselect/netstorage/tenant_filters.go:21]: // fast path
if len(tfs) == 0 {
 return tenants, tfs, nil
}

File: app/vmselect/netstorage/netstorage.go [hunk 7]
```diff
@@ -1803,6 +1927,19 @@ func processBlocks(qt *querytracer.Tracer, sns []*storageNode, denyPartialRespon
 	return isPartial, nil
 }
 
+func populateSqTenantTokensIfNeeded(qt *querytracer.Tracer, sq *storage.SearchQuery, deadline searchutils.Deadline) error {
+	if !sq.IsMultiTenant {
+		return nil
+	}
```
Comment (f41gh7) [inline app/vmselect/netstorage/netstorage.go:1933]: fast path for len(sq.TagFilterss) == 0

File: app/vmselect/netstorage/netstorage.go [hunk 8]
```diff
@@ -1803,6 +1927,19 @@ func processBlocks(qt *querytracer.Tracer, sns []*storageNode, denyPartialRespon
...
+	tts, tfss, err := GetTenantTokensFromFilters(qt, sq.GetTimeRange(), sq.TagFilterss, deadline)
+	if err != nil {
```
Comment (f41gh7) [inline app/vmselect/netstorage/netstorage.go:1935]: Apply filters to search query tokens.

File: app/vmselect/promql/eval.go [hunk 9]
```diff
@@ -1952,18 +1972,26 @@ var timeseriesByWorkerIDPool sync.Pool
 var bbPool bytesutil.ByteBufferPool
 
 func evalNumber(ec *EvalConfig, n float64) []*timeseries {
```
Comment (f41gh7) [inline app/vmselect/promql/eval.go:1974]: return previous behavior. There is no need to join authTokens.

File: app/vmselect/netstorage/netstorage.go [hunk 10]
```diff
@@ -979,7 +1004,34 @@ func LabelValues(qt *querytracer.Tracer, denyPartialResponse bool, labelName str
...
+		tenants, err := Tenants(qt, sq.GetTimeRange(), deadline)
+		if err != nil {
+			return nil, false, err
+		}
+
+		var idx int
+		switch labelName {
+		case "vm_account_id":
+			idx = 0
+		case "vm_project_id":
```
Comment (f41gh7) [inline app/vmselect/netstorage/netstorage.go:1018]: should we add default with assert:

logger.Fatalf("BUG: unexpected labeName=%q",labelName) ?

File: app/vmselect/netstorage/netstorage.go [hunk 11]
```diff
@@ -3112,3 +3251,30 @@ func (pnc *perNodeCounter) GetTotal() uint64 {
 //
 // See https://github.com/golang/go/blob/704401ffa06c60e059c9e6e4048045b4ff42530a/src/runtime/malloc.go#L11
 const maxFastAllocBlockSize = 32 * 1024
+
+// execSearchQuery calls cb for with marshaled requestData for each tenant in sq.
+func execSearchQuery(qt *querytracer.Tracer, sq *storage.SearchQuery, cb func(qt *querytracer.Tracer, requestData []byte, t storage.TenantToken) any) []any {
+	if sq.IsMultiTenant && sq.TenantTokens == nil {
+		logger.Panicf("BUG: missing TenantTokens in multi-tenant search query")
```
Comment (f41gh7) [inline app/vmselect/netstorage/netstorage.go:3258]: This assert looks a bit tricky to me.

 It's a common practice in go to use a nil slices.  And later after some refactoring TenantTokens could become nil slice.
 In this case vmselect may panic in runtime, if provided filters matches 0 tenants.

 I think, it's better to remove this assert.

File: app/vmselect/netstorage/tenant_filters.go [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+	tts, err := applyFiltersToTenantsStr(tenants, tenantFilters)
+	if err != nil {
+		return nil, nil, fmt.Errorf("cannot apply filters to tenants: %w", err)
+	}
+
+	return tts, otherFilters, nil
+}
+
+func splitFiltersByType(tfs [][]storage.TagFilter) ([]string, [][]storage.TagFilter) {
```
Comment (f41gh7) [inline app/vmselect/netstorage/tenant_filters.go:32]: I think, it's better to keep filters signature as is:

`func splitFiltersByType(tfs [][]storage.TagFilter) ([]storage.TagFilter, [][]storage.TagFilter)`

And convert TagFilter at `applyFiltersToTenants`

File: lib/storage/search.go [hunk 13]
```diff
@@ -427,6 +479,13 @@ func tagFiltersToString(tfs []TagFilter) string {
 func (sq *SearchQuery) Marshal(dst []byte) []byte {
 	dst = encoding.MarshalUint32(dst, sq.AccountID)
 	dst = encoding.MarshalUint32(dst, sq.ProjectID)
+	dst = sq.MarshaWithoutTenant(dst)
+	return dst
+}
+
+// MarshaWithoutTenant appends marshaled sq without AccountID/ProjectID to dst and returns the result.
+// It is expected that TenantToken is already marshaled to dst.
+func (sq *SearchQuery) MarshaWithoutTenant(dst []byte) []byte {
```
Comment (f41gh7) [inline lib/storage/search.go:488]: Looks like this change will break API compatibility. 

 Unmarshal requires Account and Project IDs:

```
	if len(src) < 4 {
		return src, fmt.Errorf("cannot unmarshal AccountID: too short src len: %d; must be at least %d bytes", len(src), 4)
	}
	sq.AccountID = encoding.UnmarshalUint32(src)
	src = src[4:]

	if len(src) < 4 {
		return src, fmt.Errorf("cannot unmarshal ProjectID: too short src len: %d; must be at least %d bytes", len(src), 4)
	}
```

File: lib/storage/search.go [hunk 14]
```diff
@@ -427,6 +479,13 @@ func tagFiltersToString(tfs []TagFilter) string {
 func (sq *SearchQuery) Marshal(dst []byte) []byte {
```
Comment (f41gh7) [inline lib/storage/search.go:479]: This method is only used at tests. Looks like it could be removed

**Review by f41gh7 — APPROVED** [review]
LGTM

@valyala PTAL

**Comment by jetalone85** [comment]
Hi @zekker6, @valyala and @f41gh7!

Any chance we'll see these changes in the next release? Looks almost ready, right?

**Comment by zekker6** [comment]
Hi @jetalone85, I cann't commit on that but that is very likely to happen. I'm currently working on polishin support of `/export` endpoints, but the rest is already reviewed and approved.

**Comment by zekker6** [comment]
>  Any help needed with code review?

If you will be interested, I will publish the latest version which should properly support export APIs within the next few hours. I can publish a pre-build docker images for testing if you will have some time and would like to help with identifying any issues left there.
Also, you can try multi-tenant querying on our playground here  - [link](https://play-grafana.victoriametrics.com/explore?schemaVersion=1&panes=%7B%22pcw%22:%7B%22datasource%22:%22P8FF7EDDF703E5E58%22,%22queries%22:%5B%7B%22refId%22:%22A%22,%22expr%22:%22up%22,%22range%22:true,%22instant%22:true,%22datasource%22:%7B%22type%22:%22prometheus%22,%22uid%22:%22P8FF7EDDF703E5E58%22%7D,%22editorMode%22:%22code%22,%22legendFormat%22:%22__auto%22%7D%5D,%22range%22:%7B%22from%22:%22now-1h%22,%22to%22:%22now%22%7D%7D%7D&orgId=1&chunkNotFound). You can check it out and let me know if you'll find any issues there.

**Comment by zekker6** [comment]
The latest build for this PR: [victoriametrics/vmselect:heads-multitenant-queries-0-gd022c41682](https://hub.docker.com/layers/victoriametrics/vmselect/heads-multitenant-queries-0-gd022c41682/images/sha256-765176235b919333132c79438a69d5a42c671ce859e88bafa3f71740bc35c376?context=explore)
It was also deployed to sandbox and play environments for testing.

It seems to be passing all tests for the recently discovered issues such as export API support and proper deduplication handling. @f41gh7 PTAL when you'll have some time.

File: docs/Cluster-VictoriaMetrics.md [hunk 15]
```diff
@@ -113,6 +113,44 @@ such as [Graphite](https://docs.victoriametrics.com/#how-to-send-data-from-graph
...
+```
+
+Note that `vm_account_id` and `vm_project_id` labels support all operators for label matching. For example:
+```
+up{vm_account_id!="42"} # selects all the time series except those belonging to accountID=42
+up{vm_account_id=~"4.*"} # selects all the time series belonging to accountIDs starting with 4
+```
+
 **Security considerations:** it is recommended restricting access to `multitenant` endpoints only to trusted sources,
 since untrusted source may break per-tenant data by writing unwanted samples to arbitrary tenants.
```
Comment (hagen1778) [inline docs/Cluster-VictoriaMetrics.md:155]: "or get access to data of arbitrary tenants"

File: docs/Cluster-VictoriaMetrics.md [hunk 16]
```diff
@@ -113,6 +113,44 @@ such as [Graphite](https://docs.victoriametrics.com/#how-to-send-data-from-graph
...
+For example, the following query fetches the total number of time series for the tenants `accountID=42` and `accountID=7, projectID=9`:
+```
+up{vm_account_id="7", vm_project_id="9" or vm_account_id="42"}
+```
+
+In order to achieve the same via `extra_filters[]` and `extra_label` query args, the following query must be used:
+```
+curl 'http://vmselect:8481/select/multitenant/prometheus/api/v1/query' \
+  -d 'query=up' \
+  -d 'extra_filters[]={vm_account_id="7",vm_project_id="9"}' \
```
Comment (hagen1778) [inline docs/Cluster-VictoriaMetrics.md:144]: what takes precedence here? 
Can I limit access to a subset of tenants via `extra_filters` but still allow filtering by tenant on query time? If yes, this is an important use case and should be reflected in here in the doc as well.

**Review by hagen1778 — APPROVED** [review]

File: app/vmselect/netstorage/netstorage.go [hunk 17]
```diff
@@ -1441,6 +1527,15 @@ func (tbfw *tmpBlocksFileWrapper) RegisterAndWriteBlock(mb *storage.MetricBlock,
 	m := tbfwLocal.m
 	metricName := mb.MetricName
 	addrsIdx := tbfwLocal.prevAddrsIdx
+
+	if isMultiTenant {
+		accountID := encoding.UnmarshalUint32(metricName[:4])
+		projectID := encoding.UnmarshalUint32(metricName[4:8])
```
Comment (f41gh7) [inline app/vmselect/netstorage/netstorage.go:1533]: This code is unsafe and fragile. It uses internal knowledge of storage.Metric marshal implementation. 

It's better to mutate metricName during results processing at RunParallel.

 `shouldConvertTenantToLabels` field could be added to the `Results` struct.

And be checked before calling callback at RunParallel.

File: app/vmselect/netstorage/netstorage.go [hunk 18]
```diff
@@ -1644,11 +1760,28 @@ func SearchMetricNames(qt *querytracer.Tracer, denyPartialResponse bool, sq *sto
...
+		{
+			Key:   []byte("vm_account_id"),
+			Value: strconv.AppendUint(nil, uint64(accountID), 10),
+		},
+		{
+			Key:   []byte("vm_project_id"),
+			Value: strconv.AppendUint(nil, uint64(projectID), 10),
+		},
+	}
+	suffix := make([]byte, 0)
```
Comment (f41gh7) [inline app/vmselect/netstorage/netstorage.go:1778]: It's better to rename suffix to suffixBuff and pre-allocate needed cap.

**Comment by f41gh7** [comment]
Thanks for contribution!

**Pull request merged**